numpy/__init__.pyi

```diff
@@ -1231,3 +1231,114 @@ def amin(
     initial: _NumberLike = ...,
     where: _ArrayLikeBool = ...,
 ) -> Union[number, ndarray]: ...
+
+# TODO: `np.prod()``: For object arrays `initial` does not necasarily
+# have to be a numerical scalar.
+# The only requirement is that it is compatible
+# with the `.__mul__()` method(s) of the passed array's elements.
+
+# Note that the same situation holds for all wrappers around
+# `np.ufunc.reduce`, e.g. `np.sum()` (`.__add__()`).
+
+@overload
+def prod(
+    a: _Number,
+    axis: Optional[_ShapeLike] = ...,
+    dtype: DtypeLike = ...,
+    out: None = ...,
+    keepdims: bool = ...,
+    initial: _NumberLike = ...,
+    where: _ArrayLikeBool = ...,
+) -> _Number: ...
+@overload
+def prod(
+    a: ArrayLike,
+    axis: None = ...,
+    dtype: DtypeLike = ...,
+    out: None = ...,
+    keepdims: Literal[False] = ...,
+    initial: _NumberLike = ...,
+    where: _ArrayLikeBool = ...,
+) -> number: ...
+@overload
+def prod(
+    a: ArrayLike,
+    axis: Optional[_ShapeLike] = ...,
+    dtype: DtypeLike = ...,
```

Comment:
should dtype also be optional since dtype=None is also valid ?

Reply:
As `None` is already a valid argument for `np.dtype` it is already included in the union defining `DtypeLike`:

``` python
In [1]: import numpy as np                                                                    

In [2]: np.dtype(None)                                                                        
Out[2]: dtype('float64')
```

Reply:
Yep thats true, not sure I can explain the below:
```
reveal_type(np.prod(d, dtype=None))
```
errors out as

error: No overload variant of "prod" matches argument types "ndarray", "None". Maybe also worth adding it to the tests.

Reply:
> error: No overload variant of "prod" matches argument types "ndarray", "None".

Odd, are you sure you've got the correct branch checked out?
I can't seem to reproduce the issue over here:
``` python
import numpy as np      

a = np.array(1)
b = np.prod(d, dtype=None)

reveal_type(b)
```

``` shell
>>> mypy ./test.py
test.py:6: note: Revealed type is 'numpy.number'
```

Reply:
oops, yes it was something to do with my env (had a modified test elsewhere). thanks for the clarification.